### PR TITLE
feat(clustering): directories are the structure, small ones placed by their role; 9 the target and 15 the limit at every depth

### PR DESCRIPTION
**Pull request by Svilen-Stefanov**
Stacked on #583. Five commits: the un-merge fix, the rebuild of the clustering rules around one idea: **a directory is the structure, and a small directory is placed by its role in the calls.** Replaces the earlier hub-rule commit of this PR, which produced 50 root boxes on abp. Then the review round, one naming rule for every adapter, and the end of the Unassigned bucket.

## The rules, the same at every depth

**Structure**
1. A file's position in the tree is its directory under the repository root. Names are no longer parsed for structure, so the engine root, a dotted project directory (`BTCPayServer.Client`) and each adapter's spelling stop deciding the boxes.
2. A directory holding nearly all of a scope (80%) is stepped through; every other directory is a box and its inside is the next depth. The 50% opening rule and the layout-word list are gone.

**Placement by role** (small candidates, under 5% of the scope; edge direction is what tells them apart)
3. A hub (called by 40% of its siblings), an application (calls, never called), a project root (a manifest in its directory) or a candidate shared by three siblings stays.
4. A helper, called by one sibling only, goes inside that sibling. Loose files, tests, samples, benches, docs and hubs own nothing: the bus dispatching to a service's handlers does not make the service its helper.
5. Shared by exactly two: into the larger caller. Linked to nothing: into the loose files. A hub under three files: into the loose files.

**Budget**
6. Over 9: the smallest joins the sibling carrying at least half of its links, never a hub. Over 15: the two candidates whose identifier vocabulary is closest merge. Still over 15: the smallest are pooled into one box named "Other files".
7. A folded component opens at the next depth through the same rules (the first commit).

The vocabulary rung (one candidate per word, the source of today's 51- and 60-child scopes) is deleted. Tree specifications drafted before this replay under a different position scheme, so `SPEC_VERSION` is bumped and a full run is required once.

## Measured, no model, fifteen repositories at depth 3

Root boxes · widest scope at depth 1 / 2 / 3 · scopes over 15.

| repo | today | this PR |
|---|---|---|
| eShop | 9 · 9/9/8 · 0 · **5 of 9** published boxes | 10 · 10/8/9 · 0 · **9 of 9** (F1 1.000) |
| abp | 9 · 9/**31**/9 · 4, four 1,500-file bags | 6 · 6/15/12 · 0: framework, modules, templates, npm, tools, abp_io |
| btcpayserver | **27** · 27/16/9 · 2 | 8 · 8/11/15 · 0: its eight projects |
| nopCommerce | 8 · 8/9/**51** · 1 | 4 · 4/15/15 · 0 |
| Polly | 9 · 9/8/9, strategies scattered at root | 8 · 8/11/6: its projects |
| AutoMapper | 9 · 9/7/6, test folders at root | 4 · 4/7/8 |
| serilog | 9 (forced from 13) | 9 · 9/4/4, helpers inside Core, Configuration, Formatting |
| mermaid | 9 | 6 · 6/11/10 |
| CodeBoarding | 9 | 9 · 9/4/7 |
| django, LMCache, mem0, junit4, fzf, markitdown, nanoclaw, modulify | 3 to 9 | 3 to 11, no scope over 15 |

Not one scope above 15 on any repository at any depth. abp's csproj V-measure at depth 1 reads 0.279 against today's 0.447 and that is by construction: six boxes over two hundred projects cannot be homogeneous, and the six are the repository's own top level.

## Before and after, in the product's viewer

Same graph, today's fold on the left, this PR's on the right (CodeBoarding-evals `eval/abp-type-references`, section "The fold").

eShop
abp
Polly

## After review

The bot comments were each addressed in the third commit: containment of paths under the repository root, the nearest manifest above a file as its project, file keys distinct from directory prefixes, collision-free candidate keys, one-link callers as noise, the cap on helper and vocabulary merges, consumers neither taking nor folding, fallback-only parts never folding, hubs and projects standing under a rung's floor, and the design document brought in line. Replies sit on each thread.

## Verification

- 2,155 unit tests pass (the only failure on this machine is the pre-existing macOS `/private/tmp` case in `tests/test_cli_parser.py`). 29 new tests cover positions from paths, the single frontier rule, each placement rule, the dominant-partner budget, the vocabulary merge, the pool and what the review found.
- The sweep above is the zero-LLM harness in the evals branch's README; every number there is reproducible from the pickled graphs.

## Names spelled from the repository root

One rule for every adapter: a qualified name starts with the file's path from the repository root, segment for segment as it is on disk. The C# adapter no longer drops `src`, the Java adapter no longer drops `src/main/java`, and a nested solution's or project's engine no longer spells its files relative to itself, so the same directory carries the same prefix whatever language or engine produced the symbol, and two solutions with the same inner layout cannot collide. The builder takes the repository root beside the engine root it starts the server on; the incremental path passes it through; cache tag v10. Clustering positions files by path already, so the trees on all seventeen repositories are byte-identical before and after this commit; what changes is what a reader sees in a name (`src.Basket.API.Grpc.BasketService`, `mockito-core.src.main.java.org.mockito.Answers`). The exact edge-case fixtures in CodeBoarding-tests are regenerated on the paired branch.

## No Unassigned bucket

A settled scope whose replay leaves a unit unplaced now gets a fallback-only "Loose files" rule on its own prefix, drafted on the spot, both in a full run and when an incremental run admits new units. The `Unassigned` rule kind and its bucket are gone, so a reader meets one name for the files a scope holds directly.

## What stays open

- Inside a wide flat scope (abp's `framework`, a hundred packages) the rules give fifteen package families, not the eight themes a reader would draw. That is a semantic gap; the dependency-fingerprint proxy and, failing that, the planner on such scopes only, are the next steps.
- Ocelot cannot be analysed on the stack: a nested `samples/Eureka` solution with a missing project reference raises the fatal "0 symbols" error and aborts the whole run. Separate fix.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Clustering now organizes files by repository directories and recognizes project boundaries from common manifest files.
  * Grouping better reflects caller–callee relationships, roles, shared infrastructure, vocabulary, and file-count limits.
  * Large or loosely related sets are consolidated into clearer “Other files” groupings.
  * Hierarchy generation now provides more consistent root, component, file, and fallback groupings.
  * Generated path keys use slash-separated directory paths.

* **Documentation**
  * Updated the name-tree design documentation to describe the revised file-based clustering behavior.

### CODE REVIEW

File: static_analyzer/clustering/names/draft.py [hunk 1]
```diff
@@ -180,19 +189,20 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
     def _fold(
         sources: list[int], live: list[int], sizes: list[int], links: list[list[int]], context: GroupingContext
     ) -> tuple[int, int] | None:
-        """The smallest source with an affine sibling, and that sibling: ``(source, target)``."""
-        degree = [sum(row) for row in links]
-        total = sum(degree) / 2 or 1.0
+        """The smallest source linked to a non-hub sibling, and the sibling it links to most: ``(source, target)``."""
         cap = CAP_SHARE * context.unit_count
+        threshold = max(MIN_HUB_PARTNERS, HUB_SHARE * (len(live) - 1))
+        hubs = {index for index in live if sum(1 for other in live if links[index][other] >= MIN_LINKS) >= threshold}
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Stale since the rebuild: a hub is now a fan-in notion (called by 40% of its siblings, three at least), fixed at the start of the fold, so a clique of four one-file candidates is not four hubs; the island test covering that fan passes.

File: static_analyzer/clustering/names/draft.py [hunk 2]
```diff
@@ -180,19 +189,20 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
             for target in live:
                 count = links[source][target]
-                if target == source or count < MIN_LINKS or sizes[source] + sizes[target] > cap:
+                if target == source or target in hubs or count < MIN_LINKS or sizes[source] + sizes[target] > cap:
                     continue
-                affinity = count * total / (degree[source] * degree[target])
-                if best is None or (affinity, -sizes[target], -target) > best:
-                    best = (affinity, -sizes[target], -target)
+                if best is None or (count, -sizes[target], -target) > best:
+                    best = (count, -sizes[target], -target)
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: docs/design/name-tree.md §3.1–3.3 and §3.5 now describe the rules that ship (path positions, the single frontier rule, placement by role, the raw-count fold with hubs excluded and why the ratio went), and §10 carries the remaining semantic gap.

File: static_analyzer/clustering/names/draft.py [hunk 3]
```diff
@@ -180,19 +189,20 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
-        degree = [sum(row) for row in links]
-        total = sum(degree) / 2 or 1.0
+        """The smallest source linked to a non-hub sibling, and the sibling it links to most: ``(source, target)``."""
         cap = CAP_SHARE * context.unit_count
+        threshold = max(MIN_HUB_PARTNERS, HUB_SHARE * (len(live) - 1))
+        hubs = {index for index in live if sum(1 for other in live if links[index][other] >= MIN_LINKS) >= threshold}
         for source in sources:
-            best: tuple[float, int, int] | None = None
+            if source in hubs:
+                continue
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: a group that stands by its role (a hub, an application, a project) is marked `standing` and the files rung no longer pools it under the floor; test `test_a_small_hub_or_application_stands_at_the_files_rung`. A hub under three files is sent to the loose files on purpose.

File: static_analyzer/clustering/names/draft.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Stale since the rebuild: hub status is computed once when the fold starts and a merged group inherits it, so folding two callers together cannot demote the bus.

File: static_analyzer/clustering/names/draft.py [hunk 4]
```diff
@@ -326,6 +336,35 @@ def _leaf_reason(
...
+            part.origin,
+            part.name.removesuffix(" (residual)") if part.origin == RESIDUAL else part.name,
+            part.prefixes,
+            part.fallback_prefixes,
+            part.terms,
+        )
+        for index, part in enumerate(parts)
+    ]
+    context = _context(scope_id, units, candidates, role_words, UNMERGE, links)
+    return _rules_from_groups(grouper.group(candidates, context), candidates), "structural"
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py:541]: Done: a fallback-only part (loose files, a residual) is never a fold source in any step, so an un-merge cannot absorb it into an ordinary part.

File: static_analyzer/clustering/names/inventory.py [hunk 5]
```diff
@@ -25,55 +39,60 @@ class Unit:
...
-            language,
-            tuple(sorted(names)),
-            unit_position(names, graph.delimiter),
-            unit_key(names, graph.delimiter),
+    units: list[Unit] = []
+    projects: dict[tuple[str, ...], bool] = {}
+    for file_path, names in sorted(by_file.items()):
+        if Path(file_path).is_absolute() and repo_dir is None:
+            raise ValueError(f"{file_path} is absolute and no repository root was given to position it")
+        relative = Path(normalize_repo_path(file_path, repo_dir))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/inventory.py:63]: Done: a path that is absolute outside `repo_dir` or escapes it with `..` raises `ValueError`; tests `test_a_path_outside_the_repository_root_is_refused`.

[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/inventory.py:63]: Done, see the containment check in `units_from_graph` and its test.

File: static_analyzer/clustering/names/inventory.py [hunk 6]
```diff
@@ -25,55 +39,60 @@ class Unit:
...
+        position = relative.parent.parts
+        if position == (".",):
+            position = ()
+        if position not in projects:
+            projects[position] = repo_dir is not None and _is_project_root(repo_dir / Path(*position))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/inventory.py]: Done: `Unit.project` is now the nearest directory at or above the file holding a manifest (the repository root itself excluded), and a candidate is a project when that root is one of its prefixes; test `test_the_nearest_manifest_above_a_file_names_its_project`.

[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/inventory.py]: Done, same fix as the sibling comment: nearest manifest above the file.

File: static_analyzer/clustering/names/inventory.py [hunk 7]
```diff
@@ -25,55 +39,60 @@ class Unit:
...
-            tuple(sorted(names)),
-            unit_position(names, graph.delimiter),
-            unit_key(names, graph.delimiter),
+    units: list[Unit] = []
+    projects: dict[tuple[str, ...], bool] = {}
+    for file_path, names in sorted(by_file.items()):
+        if Path(file_path).is_absolute() and repo_dir is None:
+            raise ValueError(f"{file_path} is absolute and no repository root was given to position it")
+        relative = Path(normalize_repo_path(file_path, repo_dir))
+        position = relative.parent.parts
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/inventory.py:66]: Done: candidate keys join their path on `/`, which no directory name carries, so `x/a.b` and `x/a/b` are distinct keys.

File: static_analyzer/clustering/names/draft.py [hunk 8]
```diff
@@ -125,78 +151,235 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
+            chosen = None
+            for source in sources:
+                if source in self.hubs:
+                    continue
+                degree = sum(self.links[source][other] for other in live)
+                best: tuple[int, int, int] | None = None
+                for target in live:
+                    count = self.links[source][target]
+                    if target == source or target in self.hubs or count < MIN_LINKS:
+                        continue
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py:297]: Done: a consumer (tests, samples, benches, docs, loose files) is never a fold target and never a fold source in the budget step; test `test_a_consumer_takes_nothing_over_the_budget`.

File: static_analyzer/clustering/names/draft.py [hunk 9]
```diff
@@ -125,78 +151,235 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
+                self.kept.add(index)
+            elif incoming >= MIN_LINKS and len(callers) == 1:
+                self.merge(index, next(iter(callers)))
+            elif incoming >= MIN_LINKS and len(callers) == 2:
+                # The larger caller, unless taking the helper would grow it past the cap.
+                cap = CAP_SHARE * self.context.unit_count
+                fitting = [j for j in callers if self.sizes[j] + self.sizes[index] <= cap] or list(callers)
+                self.merge(index, max(fitting, key=lambda j: (self.sizes[j], -j)))
+            elif incoming or outgoing >= MIN_LINKS:
+                self.kept.add(index)
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: the condition reads `callers or outgoing >= MIN_LINKS` where `callers` already carries only relationships of at least `MIN_LINKS`.

File: static_analyzer/clustering/names/draft.py [hunk 10]
```diff
@@ -125,78 +151,235 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
+            for a in range(len(live)):
+                for b in range(a + 1, len(live)):
+                    i, j = live[a], live[b]
+                    similarity = sum(weight * vectors[j].get(word, 0.0) for word, weight in vectors[i].items())
+                    if similarity > 0 and (best is None or similarity > best[0]):
+                        best = (similarity, i, j)
+            if best is None:
+                return
+            _, i, j = best
+            self.merge(j if self.sizes[i] >= self.sizes[j] else i, i if self.sizes[i] >= self.sizes[j] else j)
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py:328]: Done: the vocabulary merge skips a pair whose combined size exceeds the cap; test `test_the_vocabulary_merge_respects_the_cap`.

File: static_analyzer/clustering/names/draft.py [hunk 11]
```diff
@@ -125,78 +151,236 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
+        if ordered:
+            return sum(weights.get((a, b), 0) for a in left for b in right)
+        return sum(weights.get((min(a, b), max(a, b)), 0) for a in left for b in right)
+
+    def _is_consumer(self, index: int) -> bool:
+        """Loose files, tests, samples, benches and docs call the scope without being part of it."""
+        if any(self.by_key[key].kind == LOOSE for key in self.members[index]):
+            return True
+        labels = (self.by_key[key].label for key in self.members[index])
+        return any(set(stems(label)) & CONSUMER_WORDS for label in labels)
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: `_Fold` reads public methods first (`live`, `merge`, `place`, `toward_budget`, `by_vocabulary`, `pool`, `groups`), then `_between`, `_is_consumer`, `_tfidf`; `pooled` is set in `__init__`.

File: tests/static_analyzer/names/test_draft.py [hunk 12]
```diff
@@ -514,6 +540,7 @@ def context(
         FRONTIER,
         sizes={f"box:{name}": size for name, size in sizes.items()},
         links={tuple(sorted((f"box:{a}", f"box:{b}"))): count for (a, b), count in links.items()},  # type: ignore[misc]
+        calls={(f"box:{a}", f"box:{b}"): count for (a, b), count in links.items()},
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply tests/static_analyzer/names/test_draft.py]: Done: the test context sums both directions of a pair into `links`, the way `_context` does.

File: static_analyzer/clustering/names/draft.py [hunk 13]
```diff
@@ -125,78 +151,236 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
+    def place(self) -> None:
+        """Every candidate under the floor, by its role in the calls."""
+        for index in self.live():
+            if self.sizes[index] >= self.floor:
+                continue
+            # A hub calling a small sibling makes it a subscriber, not a helper: the bus owns no service.
+            callers = {
+                j: self.calls[j][index]
+                for j in self.live()
+                if j != index and j not in self.consumers and j not in self.hubs and self.calls[j][index]
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: a caller counts only with at least `MIN_LINKS` into the candidate; test `test_two_one_link_callers_are_noise_not_a_shared_helper`.

File: static_analyzer/clustering/names/draft.py [hunk 14]
```diff
@@ -125,78 +151,236 @@ def group(self, candidates: Sequence[Candidate], context: GroupingContext) -> li
...
+            outgoing = sum(self.calls[index][j] for j in self.live() if j != index)
+            if index in self.hubs:
+                if self.sizes[index] < MIN_HUB_UNITS and self.loose is not None and self.loose != index:
+                    self.merge(index, self.loose)
+                else:
+                    self.kept.add(index)
+            elif index in self.projects or len(callers) >= 3:
+                self.kept.add(index)
+            elif incoming >= MIN_LINKS and len(callers) == 1:
+                self.merge(index, next(iter(callers)))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: a helper folds into an owner only under the cap, and stays when no owner can take it; test `test_a_helper_stays_when_its_owner_cannot_take_it_under_the_cap`.

File: static_analyzer/clustering/names/inventory.py [hunk 15]
```diff
@@ -25,55 +39,60 @@ class Unit:
...
+            position = ()
+        if position not in projects:
+            projects[position] = repo_dir is not None and _is_project_root(repo_dir / Path(*position))
+        units.append(
+            Unit(
+                file_path,
+                language,
+                tuple(sorted(names)),
+                position,
+                position + (relative.stem,),
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/inventory.py]: Done: a unit's key is its position plus the file's name with its extension, so `pkg/foo.py` owns `("pkg", "foo.py")` and is not an ancestor of `pkg/foo/stray.py`; labels drop the extension.

File: static_analyzer/clustering/names/draft.py [hunk 16]
```diff
@@ -263,30 +444,26 @@ def draft_scope(
...
-        rungs.append((VOCABULARY, vocabulary))
-    elif len(scope_units) > LEAF_UNITS:
-        transpose = len(scope_units) > LEAF_CAP
-        rungs.append((SEGMENT, lambda: frontier(SEGMENT, transpose)))
-        if transpose:
-            rungs.append((VOCABULARY, vocabulary))
+        rungs.append((UNMERGE, lambda: _unmerge_rules(scope_id, scope_units, parts, role_words, grouper, links)))
+    if is_root(scope_id) or len(scope_units) > LEAF_UNITS:
+        transpose = is_root(scope_id) or len(scope_units) > LEAF_CAP
+        rungs.append((FRONTIER if is_root(scope_id) else SEGMENT, lambda: frontier(SEGMENT, transpose)))
```
[the remark this answers is not in the document]
Comment (Svilen-Stefanov) [reply static_analyzer/clustering/names/draft.py]: Done: the root passes `FRONTIER` into its grouping context; deeper scopes pass `SEGMENT`.

File: static_analyzer/clustering/names/draft.py [hunk 17]
```diff
@@ -44,22 +44,38 @@
...
 """No fold may grow a component past this share of its scope."""
-UNPLACED_NAME = "Unassigned"
 LOOSE_NAME = "Loose files"
+OTHER_NAME = "Other files"
 ISLAND_SHARE = 1 / 3
 """Share of a scope one family must hold to be drawn on its own against the rest."""
 ISLAND_STRAYS = 1
 """Links a family may exchange with the rest and still count as an island: one is noise."""
+CONSUMER_WORDS = frozenset(
+    {"test", "spec", "e2e", "sample", "example", "doc", "bench", "benchmark", "snippet", "demo", "script", "cypress"}
```
Comment (ivanmilevtues) [inline static_analyzer/clustering/names/draft.py:73]: this seems sus short. ALso Consumer words?
I don't associate  test, spec, e2e sample etc. as consumer what are they consuming?

**Review by ivanmilevtues — APPROVED** [review]
Looks good a single comment, tbh i didn't do a torough job.
